### ISSUE TITLE
Refine the release workflow

### ISSUE THREAD

**Issue by StevenJiang1110** [asterinas/asterinas#980]
# Background

## What is release

Release is essentially the process of updating the version number, which generally occurs after a relatively complete development cycle. 

## Why release

The reasons for updating the version number are as follows:

1. We need to update the content of Docker

- This update is a breaking change, meaning that the new Docker image cannot run the old code. A typical example is updating the toolchain, where some old code may no longer compile. 

- This update is not a breaking change, as the new Docker image can still run the old code. A typical example is adding new dependencies to the Docker image. 

2. We need to alert users of OSTD and OSDK to significant updates, which may or may not involve updating the Docker content.

## Current problems

The current issues are: 

- Our CI process is always failing during the release process. 

- If we are only updating the Docker content and OSTD and OSDK do not require a version number update, we should not make OSTD and OSDK aware of the version change.

# Why CI panics during release?

There were two reasons before:

Firstly, the unit tests in OSDK used to depend on the latest modifications in the main branch. Therefore, whenever modifications were made to both OSTD and OSDK, it was necessary to first submit and merge the modifications, and then redirect OSDK's dependency to the latest main branch. This issue has been fixed, and OSDK tests now only depend on local code instead of relying on updates on GitHub.

The second reason is that many tests depend on the new Docker image, but the workflow did not specify the dependencies for running the tests. As a result, the image does not exist when the tests are run. So this is the main reason.

# Proposal

## A draft version

First, let's resolve the issue with CI panicking. The main problem we encountered was the lack of dependency specification between workflows. 

For example, kernel tests and OSDK tests should have been run after the completion of the docker push. After researching, we found that GitHub provides syntax called `workflow_call`, which allows one workflow to call another workflow. In this way, the called workflow becomes a job within the calling workflow, and dependencies can be specified between multiple jobs.

Currently, the dependency graph of a CI workflow looks like this: [link to the graph](https://github.com/StevenJiang1110/asterinas/actions/runs/9706545944). 

In general, the execution order follows the dependency of the operations. Operations that cannot be undone are placed at the end, such as publishing OSDK.

Now, the release process follows these steps:

1. Submit a PR to update the version. This PR only checks if the new image can be built successfully, without running the tests.

2. After the PR is merged, it triggers operations like docker push and a series of tests. If all tests pass, it proceeds to publish the documentation and release a new version of OSDK.

4. If any tests in step 2 fail, you can continue making modifications on the main branch.

6. Once all the tests on the main branch pass, the maintainer manually triggers the release ci. This CI process completes the build image and releases OSDK to `crates.io`, along with publishing the documentation.

## The remaining problems

This operation is the simplest, but there is one issue: there are no checks when submitting a PR to update the version. These checks occur only after the code is merged. This could potentially result in problematic code being merged into the mainline. Although this won't cause severe issues since no releases are made, it is logically not ideal.

There are two possible solutions: 

- One solution is to modify the current CI logic. If a version change is detected, instead of pulling from Docker Hub, we can directly rebuild the image within the CI process. All tests will then run within this newly built image. 

- Another solution is to add a temporary `docker-push` branch.

However, this branch does not need to be synchronized with the main branch. This is mainly because keeping them in sync is challenging. For example, if an update to the image is submitted to the docker-push branch, but another PR is merged into the main branch, it becomes difficult to determine whether the changes should be synchronized with the docker-push branch. 

Therefore, I suggest that the `docker-push` branch should either be a temporary branch that gets deleted after each use or a branch that only retains code related to building the Docker image, such as `tools` and the content under `osdk/tools`. This way, the version update process becomes:

1. Submit a PR with the update to the image to the `docker-push` branch. No extensive checks are required during the PR, and the build and image push operations are triggered after the PR is merged.

2. Cherry-pick the update to the image, include other content, and submit it to the main branch. This PR can trigger tests that rely on the new image.

3. After merging into the main branch, the image build and push operations are triggered again, followed by a recheck. This ensures that changes, such as forgetting the cherry-pick or modifying other content, are addressed.
In summary:
- This operation ensures that the code merged into the mainline can also be checked during the PR.
- Have a dedicated branch for image updates, and submit code to this branch whenever an image update is required.
- Once the checks on this branch are passed, submit the code to the mainline, including the code for updating the image.

Another problem is that whether should split the test and release process. The test workflow will automatically triggerd, while the release workflow should only be manually triggered by the maintainer.

# Version Management

Another issue is related to versioning. Often, we update the Docker image to add benchmarks without the need to expose the updated version to OSDK or OSTD users. Hence, we should not update the version of OSDK or OSTD. This implies that we need to manage the Docker image version separately from the versions of OSDK and OSTD.

Generally, version numbers consist of three digits: `x.y.z`, representing major, minor, and patch versions, respectively. Currently, `x` is set to `0` for Asterinas. The proposal is to allow different values for the `z` (patch) position in OSDK/OSTD versions, while keeping `x` and `y` the same. We can modify the "bump_version" script to ensure that when updating the patch position, only the version of the image or OSDK/OSTD is modified, not both. However, this leads to an unconventional correspondence between two versions. For example, an OSDK/OSTD version of `0.5.1` would correspond to the latest image version of `0.5`. Please note that `0.5` represents the latest image version within the `0.5.z` range.

## Examples

Let's illustrate this with a specific workflow. 

Assume the current Docker image version is `0.5.1`. 

If we want to make a non-breaking change to the Docker image, such as adding a dependency library or benchmark, and the modified image can still run the existing code, we can create a PR to update the Docker image version to `0.5.2`. The CI pipeline will automatically tag and push this image with both `0.5` and `0.5.2`. 

If we need to make a breaking change to the Docker image, such as updating the toolchain, the update should require a change in the minor version. In this case, the PR should update the OSDK/OSTD version to `0.6.0`, and the image version should also be updated to `0.6.0/0.6`, simultaneously having two labels. This modification must be synchronized with the release of OSDK and OSTD.

## Summary
1. Allow the patch position to be different between the version of the Docker image and OSDK/OSTD.
5. OSDK/OSTD/Asterinas with version `x.y.z` can always run in a container with version `x.y`.

**Comment by lrh2000** [comment]
> This operation is the simplest, but there is one issue: there are no checks when submitting a PR to update the version. These checks occur only after the code is merged.

Could [merge queues](
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue) help solve this problem?

**Comment by StevenJiang1110** [comment]
> Could [merge queues](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue) help solve this problem?

Thanks for your advice!

I briefly looked into the docs of the merge queue. The merge queue tries to create a new branch and perform checks when maintainer attempts to merge a pull request. If it fails, the merge queue is popped out. This provides us with a new checkpoint, where the maintainer agrees to merge but it hasn't been merged into the main branch yet. This way, there is no need to create a new branch `docker-push`, and this approach seems much more elegant.

I feel that combining the merge queue with `workflow_call` can fully solve the CI failure issue. I need to further experiment with it.
